File: day_2/2.py
import queue

sum = 0

def is_invalid_id(s: str) -> bool:
    return s in (s[:i] * (len(s) // i) for i in range(1, len(s)//2 + 1) if len(s) % i == 0)

with open("2.dat", "r") as f:
    for ranges in f.readlines()[0].split(","):
        start, end = ranges.split("-")
        for number in range(int(start), int(end) + 1):
            if is_invalid_id(str(number)):
                sum += number

    print(sum)



            

File: day_2/2.hs
main :: IO ()
main = do
    contents <- readFile "2.dat"
    let firstLine = head (lines contents)
        ranges    = splitBy ',' firstLine
        nums      = concatMap expandRange ranges
        goodNums  = filter isInvalidId nums
    print (sum goodNums)

splitBy :: Char -> String -> [String]
splitBy _ ""     = [""]
splitBy d (c:cs)
    | c == d    = "" : rest
    | otherwise = (c : head rest) : tail rest
  where
    rest = splitBy d cs

expandRange :: String -> [Int]
expandRange s =
    let [a,b] = splitBy '-' s
    in [read a .. read b]

isInvalidId :: Int -> Bool
isInvalidId n =
    let s   = show n
        len = length s
        -- i runs from 1 to len/2, and must divide len
        possible i = len `mod` i == 0 &&
                     concat (replicate (len `div` i) (take i s)) == s
    in any possible [1 .. len `div` 2]
